### PR TITLE
Correct Logging for Black Oil Well Model for parallel runs

### PR DESCRIPTION
**Pull request by fgfuchs**
use deferred logging to log correctly in parallel for wells

### CODE REVIEW

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 1]
```diff
@@ -876,9 +876,13 @@ namespace Opm {
         } while (it < max_iter);
 
         if (converged) {
-            deferred_logger.debug("Well equation solution gets converged with " + std::to_string(it) + " iterations");
+            if (terminal_output_) {
```
Comment (blattms) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: I guess you need to change the initialization of `terminal_output_` to achieve logging on all processes. Currently it is initialized by
```
terminal_output_ = false;
if (ebosSimulator.gridView().comm().rank() == 0)
  terminal_output_ = EWOMS_GET_PARAM(TypeTag, bool, EnableTerminalOutput);
```

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 2]
```diff
@@ -311,36 +311,35 @@ namespace Opm {
     template<typename TypeTag>
     void
     BlackoilWellModel<TypeTag>::wellTesting(const int timeStepIdx, const double simulationTime) {
+        Opm::DeferredLogger local_deferredLogger;
         const auto& wtest_config = schedule().wtestConfig(timeStepIdx);
-        if (wtest_config.size() == 0) { // there is no WTEST request
```
Comment (blattms) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: Did we deadlock previously or what it is this reason for this change?

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 3]
```diff
@@ -386,9 +387,15 @@ namespace Opm {
         catch ( std::runtime_error& e )
         {
             const std::string msg = "A zero well potential is returned for output purposes. ";
-            OpmLog::warning("WELL_POTENTIAL_CALCULATION_FAILED", msg);
+            local_deferredLogger.warning("WELL_POTENTIAL_CALCULATION_FAILED", msg);
         }
         previous_well_state_ = well_state_;
+
+        Opm::DeferredLogger global_deferredLogger = gatherDeferredLogger(local_deferredLogger);
+        if (terminal_output_) {
```
Comment (blattms) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: If you change `terminal_output_` as suggested elsewhere this check will be true on all processes. And by design of `gatherDeferredLogger` (which uses MPI_Allgatherv) all processes will have all messages and log them. Hence multiples of all messages.

Maybe gatherDeferredLogger should be refactored to gather the messages only to one process?

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 4]
```diff
@@ -641,20 +649,26 @@ namespace Opm {
...
             }
             // TODO: should we update the explicit related here again, or even prepareTimeStep().
             // basically, this is a more updated state from the solveWellEq based on fixed
             // reservoir state, will tihs be a better place to inialize the explict information?
         }
-        assembleWellEq(dt);
+        assembleWellEq(dt, local_deferredLogger);
+
+        Opm::DeferredLogger global_deferredLogger = gatherDeferredLogger(local_deferredLogger);
+        if (terminal_output_) {
```
Comment (blattms) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: ditto

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 5]
```diff
@@ -905,11 +919,13 @@ namespace Opm {
         ConvergenceReport report = gatherConvergenceReport(local_report);
 
         // Log debug messages for NaN or too large residuals.
-        for (const auto& f : report.wellFailures()) {
-            if (f.severity() == ConvergenceReport::Severity::NotANumber) {
-                OpmLog::debug("NaN residual found with phase " + std::to_string(f.phase()) + " for well " + f.wellName());
-            } else if (f.severity() == ConvergenceReport::Severity::TooLarge) {
-                OpmLog::debug("Too large residual found with phase " + std::to_string(f.phase()) + " for well " + f.wellName());
+        if (terminal_output_) {
```
Comment (blattms) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: Mind the initialization of `terminal_output_`, see above.

Comment (blattms) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: Well, here the old initialization would be correct: Log only on one process.

**Comment by fgfuchs** [comment]
> We also need to be aware that in case any method call between the initialization of `DeferredLogger` and the call to `gatherDeferredLogger` throws an uncaught exception, we will loose all the log messages. Not sure whether this might happen, though.

Good point, I do not know how to handle that.

**Comment by fgfuchs** [comment]
> BTW: Title of PR should be changed, the commit squashed and commit message should contain a bit more detailed explanation.

Hope this is good enough

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (fgfuchs) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: > I guess we have to have another variable apart from `terminal_output` to make this work as expected and/or refactor `gatherDeferredLogger`

@blattms thank you for the comment. The way parallel logging works, is that all processes get the messages of all processes. terminal_output can be true only on rank 0. Hence, the messages from all processes get logged only once. Please, feel free to correct me, if I am wrong.

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 2 above)
Comment (fgfuchs) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: If I understand correctly, than, before this change, logging happened only from rank 0. Now we use if(wtest_config.size() != 0) to test whether we need to execute well testing, and run gatherDeferredLogger afterwards in any case.

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 3 above)
Comment (fgfuchs) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: see other comments

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 4 above)
Comment (fgfuchs) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: ditto

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 5 above)
Comment (fgfuchs) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: see my other comments

**Comment by blattms** [comment]
Well in a parallel run your PR clearly is an improvement (we did not see any logging despite of rank 0 before).

The question is how does the deferred logger work when there is only one rank? Does ist still only log messages in
the logMessages call? If so, maybe we can change this to use direct logging in case of only rank. Hence the situation
will be the same as before for a serial run. We do not loose anything but win something in a parallel run.


Great work BTW.

**Comment by atgeirr** [comment]
> If so, maybe we can change this to use direct logging in case of only rank

That may be a good idea, and I am open to it. On the other hand it could make it harder to spot cases of user-unfriendly logging (for example, if logging happens too late to make sense of it) if behaviour is different between serial and parallel. Eating the parallel dogfood in serial if you will...

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 2 above)
Comment (blattms) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: I see. Makes sense.

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (blattms) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: You are right. I oversaw that convergence is checked globally.

**Comment by blattms** [comment]
With a good night of sleep and glasses on my nose. This now looks good. Sorry for bogus nitpicks before.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
Thanks @fgfuchs for your work, and @blattms for your review. All is green so I'll merge!

**Pull request merged**